HopsanGenerator/templates/fmu_hopsan.cpp

```diff
@@ -138,45 +148,86 @@ void hopsan_get_message(hopsan_message_callback_t message_callback, void* userSt
 
 double hopsan_get_real(int vr)
 {
-    return (*dataPtrs[vr]);
+    if(vr < <<<nports>>>) {
+        return (*dataPtrs[vr]);
+    }
+    if(realParametersMap.count(vr)) {
+        hopsan::HString value;
+        spCoreComponentSystem->getParameterValue(realParametersMap[vr],value);
+        bool ok;
+        return value.toDouble(&ok);
+    }
+    return -1;
 }
 
 int hopsan_get_integer(int vr)
 {
-    //Write code here
+    if(intParametersMap.count(vr)) {
+        hopsan::HString value;
+        spCoreComponentSystem->getParameterValue(intParametersMap[vr],value);
+        bool ok;
+        return value.toLongInt(&ok);
+    }
     return -1;
 }
 
 int hopsan_get_boolean(int vr)
 {
-    //Write code here
+    if(boolParametersMap.count(vr)) {
+        hopsan::HString value;
+        spCoreComponentSystem->getParameterValue(boolParametersMap[vr],value);
+        bool ok;
+        return value.toBool(&ok);
+    }
     return false;
 }
 
 const char* hopsan_get_string(int vr)
 {
-    //Write code here
+    if(stringParametersMap.count(vr)) {
+        hopsan::HString value;
+        spCoreComponentSystem->getParameterValue(stringParametersMap[vr],value);
+        bool ok;
+        return value.c_str();
```

Comment:
I wonder how safe this is, since value is a temporary. The returned char * may be dangling. Maybe its OK if it is used emediately but I am not sure.

Reply:
That should not be safe. I guess it works if you're lucky and the memory has not yet been overwritten, but we should fix it. We can't return string or hstring since the function must return a C compatible data type. Do you think a strcpy to a new const char buffer before returning will work?

Reply:
Yes this is the only way, to have an internal persistent variable to return from. This could be a HString that we assign and we just return HString.c_str(), but the problem is that the function will not become thread safe, but maybe there is no such guarantee anyway. 

So you can only call this function sequentially, and you must use the returned char* instantly and copy the value into a variable. I dont know if the FMI standard says anything about this.
Well actually you can use the returned pointer as long as our internal HString value is not being changed, but that will happen when the next call to the function is made.

But its the only solution I can see.

Reply:
According to FMI specifications (2.0.1):
> The strings returned by fmi2GetString must be copied in the target environment because
the allocated memory for these strings might be deallocated by the next call to any of the
fmi2 interface functions or it might be an internal string buffer that is reused.

So if I understand correctly, the correct solution would be to make `value` a global variable to make sure it is persistent between each call to `hopsan_get_string`?

Reply:
Yes sounds correct

Reply:
Now it should work. I tested it with the unit test model, but it did of course work already before.